### PR TITLE
First commit to adding testng support

### PR DESCRIPTION
**Pull request by jossjacobo**
- TestNG Support
- ThreadSafe access to WebDriver to run tests in Parallel

### CODE REVIEW

File: src/main/java/io/ddavison/conductor/Locomotive.java [hunk 1]
```diff
@@ -24,74 +27,63 @@
...
 
-public class Locomotive implements Conductor<Locomotive> {
+@Listeners({TestListener.class, WebDriverListener.class})
+public class Locomotive extends Watchman implements Conductor<Locomotive> {
+
+    //Declare ThreadLocal Driver (ThreadLocalMap) for ThreadSafe Tests
+    private static ThreadLocal<WebDriver> driver = new ThreadLocal<>();
 
     public ConductorConfig configuration;
-    public WebDriver driver;
```
Comment (konwinkler) [inline src/main/java/io/ddavison/conductor/Locomotive.java]: I think changing the driver access to a getter will break a lot of projects but is a really good change. I think we should make clear in the release notes that the public access got removed and might require some code adjustments.

File: src/main/java/io/ddavison/conductor/Locomotive.java [hunk 2]
```diff
@@ -24,74 +27,63 @@
...
-    public void teardown() {
-        if (!configuration.isScreenshotOnFail()) {
-            driver.quit();
+    /**
+     * The listener classes ({@link Watchman}, {@link TestListener} {@link WebDriverListener}) will call this
+     * method upon test completion.
+     */
+    public void quit() {
+        if (getDriver() != null) {
+            getDriver().quit();
```
[the remark this answers is not in the document]
Comment (jossjacobo) [reply src/main/java/io/ddavison/conductor/Locomotive.java]: Whoops, yeah nice catch! Had that in there but I don't know where it went...

**Comment by konwinkler** [comment]
Looks like the tests are not executed for the PR on our CI server: http://builds.willowtreeapps.com/viewLog.html?buildId=330426&buildTypeId=WillowTreeApps_Conductor_ConductorWeb_ConductorWebPullRequestBuilds&tab=buildLog

File: src/main/java/io/ddavison/conductor/Locomotive.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jossjacobo) [reply src/main/java/io/ddavison/conductor/Locomotive.java]: Done.

**Comment by jossjacobo** [comment]
Yeah we should port this over to `conductor-mobile` there are some quality of life updates I made to this for running your tests using TestNG plus the threadsafe driver.

**Comment by jossjacobo** [comment]
@konwinkler I'll update the test runners, Intellij just runs them for me because it gets me and knows what I mean, `mvn test` doesn't seem to be as understanding hahaha.

File: src/main/java/io/ddavison/conductor/listeners/Watchman.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    public abstract Locomotive getLocomotive();
+    public abstract void quit();
+
+    private boolean failure;
+    private Throwable e;
+
+    @Override
+    protected void failed(Throwable e, Description description) {
+        if (getLocomotive().configuration.isScreenshotOnFail()) {
+            failure = true;
```
Comment (konwinkler) [inline src/main/java/io/ddavison/conductor/listeners/Watchman.java]: Why are we setting a flag here and not taking a screenshot directly here? The call to `quit()` could stay in `finished()` if needed.

Comment (jossjacobo) [reply src/main/java/io/ddavison/conductor/listeners/Watchman.java]: That's a good question, I forget way, I know there was a reason. Let me check if I can move it.

**Comment by jossjacobo** [comment]
@konwinkler they are required to run and pass, but since it didn't pick up any tests it technically was a successful run because nothing failed (even though nothing ran).

Should be fixed now, all tests (some in parallel) ran and passed in TC. 🥇

File: src/main/java/io/ddavison/conductor/util/ScreenShotUtil.java [hunk 4]
```diff
@@ -26,10 +26,12 @@ public static void take(Locomotive locomotive, String path, String testName) {
...
     private static void writeFile(TakesScreenshot takesScreenshot, String filePathAndName) {
-        try {
-            FileUtils.copyFile(takesScreenshot.getScreenshotAs(OutputType.FILE), new File(filePathAndName));
-        } catch (IOException e) {
-            e.printStackTrace();
+        if (takesScreenshot != null && filePathAndName != null) {
+            try {
+                FileUtils.copyFile(takesScreenshot.getScreenshotAs(OutputType.FILE), new File(filePathAndName));
+            } catch (IOException e) {
+                e.printStackTrace();
```
Comment (konwinkler) [inline src/main/java/io/ddavison/conductor/util/ScreenShotUtil.java]: can we do this through a logging library, like `Logger.error(e)`?

Comment (jossjacobo) [reply src/main/java/io/ddavison/conductor/util/ScreenShotUtil.java]: Done.

**Pull request merged**